policy/simulated/options.go

```diff
@@ -0,0 +1,235 @@
+}
+
+func NewOptionsFromRequest(r *http.Request) (Options, error) {
+	var o Options
+	if r.Body == nil {
+		return o, nil
+	}
+
+	if err := json.NewDecoder(r.Body).Decode(&o); err != nil {
+		return o, errors.Wrap(err, "failed to unmarshal body into options")
+	}
+
+	o.setDefaults()
+	return o, nil
+}
+
+// setDefaults sets any values for the options that were not intentionally set in the request body but which should have
+// consistent values for the length of the simulation, such as the created time for a comment or review.
+func (o *Options) setDefaults() {
+	for i, review := range o.AddReviews {
+		review.setDefaults()
+		o.AddReviews[i] = review
+	}
+
+	for i, comment := range o.AddComments {
+		comment.setDefaults()
+		o.AddComments[i] = comment
+	}
+}
+
+func (o *Options) filterIgnoredComments(ctx context.Context, prCtx pull.Context, comments []*pull.Comment) ([]*pull.Comment, error) {
+	if o.IgnoreComments == nil {
+		return comments, nil
+	}
+
+	var filteredComments []*pull.Comment
+	actors, err := o.IgnoreComments.toCommonActors()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, comment := range comments {
+		isActor, err := actors.IsActor(ctx, prCtx, comment.Author)
+		if err != nil {
+			return nil, err
+		}
+
+		if isActor {
+			continue
+		}
+
+		filteredComments = append(filteredComments, comment)
+	}
+
+	return filteredComments, nil
+}
+
+func (o *Options) filterIgnoredReviews(ctx context.Context, prCtx pull.Context, reviews []*pull.Review) ([]*pull.Review, error) {
+	if o.IgnoreReviews == nil {
+		return reviews, nil
+	}
+
+	var filteredReviews []*pull.Review
+	actors, err := o.IgnoreReviews.toCommonActors()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, review := range reviews {
+		isActor, err := actors.IsActor(ctx, prCtx, review.Author)
+		if err != nil {
+			return nil, err
+		}
+
+		if isActor {
+			continue
+		}
+
+		filteredReviews = append(filteredReviews, review)
+	}
+
+	return filteredReviews, nil
+}
+
+func (o *Options) addApprovalComment(comments []*pull.Comment) []*pull.Comment {
+	var commentsToAdd []*pull.Comment
+	for _, comment := range o.AddComments {
+		commentsToAdd = append(commentsToAdd, comment.toPullComment())
+	}
+
+	return append(comments, commentsToAdd...)
+}
+
+func (o *Options) addApprovalReview(reviews []*pull.Review) []*pull.Review {
+	var reviewsToAdd []*pull.Review
+	for i, review := range o.AddReviews {
+		reviewID := fmt.Sprintf("simulated-reviewID-%d", i)
+		reviewSHA := fmt.Sprintf("simulated-reviewSHA-%d", i)
+
+		reviewsToAdd = append(reviewsToAdd, review.toPullReview(reviewID, reviewSHA))
+	}
+
+	return append(reviews, reviewsToAdd...)
+}
+
+func (o *Options) branches(base, head string) (string, string) {
+	if o.BaseBranch != "" {
+		base = o.BaseBranch
+	}
+
+	return base, head
+}
+
+type Actors struct {
+	Users         []string `json:"users"`
+	Teams         []string `json:"teams"`
+	Organizations []string `json:"organizations"`
+	Permissions   []string `json:"permissions"`
+}
```

Comment:
Yeah I considered that, but then `permissions` on `common.Actors` is defined as a `[]pull.Permission` which is a `uint8` and I thought it would be more friendly for the request body to reference the permissions by their string value and convert them into the `uint8` during the conversion function.

Reply:
Ah, `pull.Permission` defines a custom `UnmarshalText` function so that you can use strings. While it's apparently missing the `yaml` tag, the strings work fine when used in YAML configuration and the `encoding/json` package will also use `UnmarshalText` when it is defined.

Reply:
Oh nice, I missed that. I moved the `options` to use the built in structs and added the necessary json tags.